entr.1
.\"
.\" Copyright (c) 2012 Eric Radman <ericshane@eradman.com>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd August 15, 2023
.Dt ENTR 1
.Os
.Sh NAME
.Nm entr
.Nd run arbitrary commands when files change
.Sh SYNOPSIS
.Nm
.Op Fl acdnprsz
.Ar utility
.Op Ar argument /_ ...
.Sh DESCRIPTION
A list of files provided on standard input, and the
.Ar utility
is executed using the supplied arguments if any of them change.
.Nm
waits for the child process to finish before responding to subsequent file
system events.
A TTY is also opened before entering the watch loop in order to support
interactive utilities.
.Pp
The arguments are as follows:
.Bl -tag -width Ds
.It Fl a
Respond to all events which occur while the
.Ar utility
is running.
Without this option,
.Nm
consolidates events in order to avoid looping.
This option has no effect in conjunction with the
.Fl r
flag.
.It Fl c
Clear the screen before invoking the
.Ar utility
specified on the command line.
Specify twice to erase the scrollback buffer.
.It Fl d
Track the directories of regular files provided as input and exit if a new file
is added.
This option also enables directories to be specified explicitly.
If specified twice, all new entries to a directory are recognized,
otherwise files with names beginning with
.Ql \&.
are ignored.
.It Fl n
Run in non-interactive mode.
In this mode
.Nm entr
does not attempt to read from the TTY or change its properties.
.It Fl p
Postpone the first execution of the
.Ar utility
until a file is modified.
.It Fl r
Reload a persistent child process.
As with the standard mode of operation, a
.Ar utility
which terminates is not executed again until a file system or keyboard event is
processed.
.Dv SIGTERM
is used to terminate the
.Ar utility
before it is restarted.
A process group is created to prevent shell scripts from masking signals.
.Nm
waits for the
.Ar utility
to exit to ensure that resources such as sockets have been closed.
Control of the TTY is not transferred to the child process.
.It Fl s
Evaluate the first argument using the interpreter specified by the
.Ev SHELL
environment variable.
If standard output is a TTY, the name of the shell and exit code is printed
after each invocation.
.It Fl z
Exit after the
.Ar utility
completes.
When combined with
.Fl r
the
.Ar utility
will be restarted again only in response to commands or file system
events.
.It Fl t Ar timeout
An option for restart mode.
Send
.Dv SIGKILL
if the utility does not to exit in the specified number of seconds.
.El
.Pp
The first argument named
.Ar /_
is replaced with the absolute path of the first file to trigger an event.
The first file under watch is used as the default.
If the
.Fl s
option is used, the name of the first file to trigger an event can be read from
.Va $0 .
.Sh COMMANDS
.Nm
listens for keyboard input and responds to the following commands:
.Bl -tag -width <space>
.It Aq Cm space
Execute the utility immediately.
If the
.Fl Cm r
option is set this will terminate and restart the child process as if a file
change event had occurred.
.It Cm q
Quit; equivalent pressing
.Aq Cm control-C .
.El
.Sh ENVIRONMENT
.Bl -tag -width "ENTR_ENVIRON"
.It Ev PAGER
Set to
.Pa /bin/cat
by default
to prevent interactive utilities from waiting for keyboard input if output does
not fit on the screen.
.It Ev SHELL
Specify the shell to use with the
.Fl s
flag.
The default is
.Pa /bin/sh .
.It Ev EV_TRACE
Print file system event messages.
.El
.Sh EXIT STATUS
If the
.Fl z
flag is set and the
.Ar utility
is successfully executed, the status of the child process is returned.
If the child process was terminated by a signal, the exit status is the signal
number plus 128.
.Pp
.Nm
normally returns one of the following values:
.Pp
.Bl -tag -width Ds -offset indent -compact
.It 0
Normal termination after receiving
.Dv SIGINT
.It 1
No regular files were provided as input or an error occurred
.It 2
A file was added to a directory and the directory watch option was specified
.El
.Sh EXAMPLES
Rebuild a project if source files change, limiting output to the first 20 lines:
.Pp
.Dl $ find src/ | entr -s 'make | head -n 20'
.Pp
Launch and auto-reload a node.js server:
.Pp
.Dl $ ls *.js | entr -r node app.js
.Pp
Clear the screen and run a query after the SQL script is updated:
.Pp
.Dl $ echo my.sql | entr -cp psql -f /_
.Pp
Rebuild project if a source file is modified or added to the src/ directory:
.Pp
.Dl $ while sleep 0.1; do ls src/*.rb | entr -d make; done
.Pp
Auto-reload a web server, or terminate if the server exits
.Pp
.Dl $ ls * | entr -rz ./httpd


NEWS
= Next Release: 5.5

 - Allow signal for restarting the utility to be specified as the last argument

= Release History

== 5.4: August 1, 2023

 - 'make test' runs a quick smoketest, 'make check' runs regressions
 - Set IN_CLOEXEC only for inotify_init, kqueue uses similar setting by default
 - Unconditionally try to set soft file limit to 2^16 on MacOS
 - Use non-reentrant calls sparingly in signal handlers
 - configure: use TARGET_OS to override the output of uname(1)

== 5.3: January 20, 2023

 - Symlink changes detected on Linux by setting 'ENTR_INOTIFY_SYMLINK'
 - Use /dev/null rather then closed pipe for stdin in -r mode
 - Utilize {O,FD}_CLOEXEC flag for unintentional leaks of descriptors to
   executed utilities
 - Remove C unit tests
 - Only respond to attribute/inode changes on Linux

== 5.2: May 05, 2022

 - Update copy of strlcpy(3) for Linux
 - Detect file deletion from directories on Linux
 - Print the signal that terminated a child when using '-s'
 - Return 128+signal that terminated a child when using '-z'
 - Ensure terminal settings are reset when '-z' is set

== 5.1: December 28, 2021

 - Detect files moved to or from directories on Linux
 - Allow detection of directory entries beginning with '.' by specifying '-d' twice
 - Only reset terminal settings in exit handler if settings were changed

== 5.0: July 2, 2021

 - Add missing call to getrlimit(3) on MacOS
 - Eliminate memory management warnings on Linux
 - EV_TRACE prints file mode and file name
 - Ignore asynchronous attribute changes on MacOS 11

== 4.9: May 3, 2021

 - EV_TRACE also prints file/notify descriptor limit
 - Don't raise rlim_cur on MacOS
 - Set 2^16 watches if inotify limits cannot be read
 - Raise an error and suggest '-n' if terminal attributes cannot be read

== 4.8: February 26, 2021

 - Set a maximum of 2^19 watches to guard against absurd file open limits on
   MacOS
 - Use control sequences to clear the display and specify '-c' twice to erase
   the scrollback buffer

== 4.7: January 29, 2021

 - Use system file descriptor limits when max_user_watches is not accessible
 - Return the exit status of the child process when the '-z' option is used
 - Handle SIGHUP so child process are terminated when a terminal is closed
 - More accurately return shell exit code using '-s' option

== 4.6: July 1, 2020

 - Always call waitpid(2) to avoid dead processes
 - Duplicate STDIN file descriptor before closing; for the '-r' option

== 4.5: April 20, 2020

 - New '-z' "one-shot" option self-terminates after the utility exits
 - Termination by 'q' or 'SIGINT' results in an exit status of 0

== 4.4: January 02, 2020

 - Use a single inotify queue on Linux, limited by
   /proc/sys/fs/inotify/max_user_watches
 - Set the environment variable 'ENTR_INOTIFY_WORKAROUND' to enable a
   compatibility mode for platforms with deformed inotify support

== 4.3: September 25, 2019

 - Move hosting from bitbucket.org to github.com
 - Rename tag names from entr-X.Y to X.Y
 - Replace hg with git as a dependency for system-test.sh

== 4.2: March 28, 2019

 - Use pledge(2) on OpenBSD
 - New '-a' option enables response to events that occur while the utility is
   running
 - Error messages direct users to http://eradman.com/entrproject/limits.html
 - Correctly report error when a file cannot be reopened

== 4.1: June 13, 2018

 - New '-n' non-interactive option disables keyboard input
 - EV_TRACE environment variable enables file system event tracing
 - Track changes to the inode number as a workaround for missing delete
   events on the Linux kernel

== 4.0: February 02, 2018

 - Warn instead of error if kqueue fails to register on STDIN
 - Close STDIN before running the utility when the restart option is used
 - Restore terminal settings if terminated by a signal

== 3.9: September 19, 2017

 - Fix use of poll(2) to avoid possible busy-loop on Linux
 - Disable keyboard input if reading STDIN fails

== 3.8: August 11, 2017

 - Run the utility if the spacebar is pressed
 - 'q' for quit

== 3.7: February 27, 2017

 - Terminate subprocess in restart mode if a file under watch disappears
 - Allow NOTE_ATTRIB to set '/_' only if file mode changes
 - New '-s' option executes commands using $SHELL -c
 - Print usage and exit if input is from a terminal instead of a pipe

== 3.6: July 01, 2016

 - Do not print warning if _TTY_PATH cannot be opened (for chroot, docker, ...)
 - Non-existent input files produce a warning instead of an error
 - Trigger run when permissions change on regular files

== 3.5: April 01, 2016

 - Removed contrib/*
 - Detect new subdirectories on Linux
 - Direct users to http://entrproject.org/limits.html if inotify hits a kernel
   limit

== 3.4: December 12, 2015

 - Terminate the child process before exiting when the reload option is combined
   with the directory watch option
 - Direct users to http://entrproject.org/limits.html if the maximum number of
   open files is exceeded

== 3.3: October 23, 2015

 - Substitute '/_' only with regular files when using the directory watch option
 - Bugfix when tracking the first file to change
 - Display release number along with usage


== 3.2: April 9, 2015

 - Execute the utility immediately after registering watch events. Old behavior
   is restored using new postpone option ('-p')
 - Relocated 'reload-browser' to http://entrproject.org/scripts/reload-browser
 - Drop FIFO support to further simply code and documentation. FIFO mode was
   conceptually good, but EXEC mode is intelligent hand easily handles the vast
   majority of use cases
 - Remove kFreeBSD as a target platform

== 3.1: March 4, 2015

 - Return error code indicating exit reason after handling signals

entr.c
/*
 * Copyright (c) 2012 Eric Radman <ericshane@eradman.com>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <sys/param.h>
#include <sys/resource.h>
#include <sys/wait.h>
#include <sys/stat.h>
#include <sys/event.h>

#include <dirent.h>
#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <libgen.h>
#include <limits.h>
#include <paths.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <termios.h>
#include <time.h>
#include <unistd.h>

#include "missing/compat.h"

#include "data.h"

/* events to watch for */

#define NOTE_ALL NOTE_DELETE|NOTE_WRITE|NOTE_RENAME|NOTE_TRUNCATE|NOTE_ATTRIB

/* shortcuts */

#define min(a, b) (((a) < (b)) ? (a) : (b))
#define MEMBER_SIZE(S, M) sizeof(((S *)0)->M)

/* globals */

extern int optind;
WatchFile **files;
WatchFile *leading_edge;
int child_pid;
int child_status;
int terminating;

int aggressive_opt;
int clear_opt;
int dirwatch_opt;
int noninteractive_opt;
int oneshot_opt;
int postpone_opt;
int restart_opt;
int shell_opt;
int timeout_arg;

int termios_set;
struct termios canonical_tty;

static char *shell, *shell_base;

/* forwards */

static void usage();
static void terminate_utility();
static void handle_exit(int sig);
static void proc_exit(int sig);
static void print_child_status(int status);
static int process_input(FILE *, WatchFile *[], int);
static int set_options(char *[]);
static int list_dir(char *);
static void run_utility(char *[]);
static void watch_file(int, WatchFile *);
static int compare_dir_contents(WatchFile *);
static void watch_loop(int, char *[]);

/*
 * The Event Notify Test Runner
 * run arbitrary commands when files change
 */
int
main(int argc, char *argv[]) {
	struct rlimit rl;
	int kq;
	struct sigaction act;
	int ttyfd;
	short argv_index;
	int n_files;
	int i;
	struct kevent evSet;
	unsigned open_max;

	 if (pledge("stdio rpath tty proc exec", NULL) == -1)
	    err(1, "pledge");

	/* call usage() if no command is supplied */
	if (argc < 2) usage();
	argv_index = set_options(argv);

	sigemptyset(&act.sa_mask);

	/* normally a user will exit this utility by do_execting Ctrl-C */
	act.sa_flags = SA_RESETHAND;
	act.sa_handler = handle_exit;
	if (sigaction(SIGINT, &act, NULL) != 0)
		err(1, "Failed to set SIGINT handler");
	if (sigaction(SIGTERM, &act, NULL) != 0)
		err(1, "Failed to set SIGTERM handler");
	if (sigaction(SIGHUP, &act, NULL) != 0)
		err(1, "Failed to set SIGHUP handler");

	/* notification used to combine the one-shot and restart options */
	act.sa_flags = 0;
	act.sa_handler = proc_exit;
	if (sigaction(SIGCHLD, &act, NULL) != 0)
		err(1, "Failed to set SIGCHLD handler");

#if defined(_LINUX_PORT)
	/* attempt to read inotify limits */
	open_max = (unsigned)fs_sysctl(INOTIFY_MAX_USER_WATCHES);
	if (open_max == 0)
		open_max = 65536;
#elif defined(_MACOS_PORT)
	if (getrlimit(RLIMIT_NOFILE, &rl) == -1)
		err(1, "getrlimit");
	/* guard against unrealistic replies */
	open_max = min(65536, (unsigned)rl.rlim_cur);
	/* maxfiles is ulimited on modern MacOS */
	if (open_max < 65536) {
		rl.rlim_cur = (rlim_t)65536;
		if (setrlimit(RLIMIT_NOFILE, &rl) == 0 || open_max == 0)
			open_max = 65536;
	}
#else /* BSD */
	if (getrlimit(RLIMIT_NOFILE, &rl) == -1)
		err(1, "getrlimit");
	open_max = (unsigned)rl.rlim_max;
	rl.rlim_cur = (rlim_t)open_max;
	if (setrlimit(RLIMIT_NOFILE, &rl) != 0)
		err(1, "setrlimit cannot set rlim_cur to %u", open_max);
#endif

	if (getenv("EV_TRACE"))
		fprintf(stderr, "open_max: %d\n", open_max);

	/* prevent interactive utilities from paging output */
	setenv("PAGER", "/bin/cat", 0);

	/* ensure a shell is available to use */
	setenv("SHELL", "/bin/sh", 0);

	shell = getenv("SHELL");
	shell_base = strdup(shell);
	if (shell_base == NULL)
		err(1, "cannot duplicate string");
	shell_base = basename(shell_base);

	/* sequential scan may depend on a 0 at the end */
	files = calloc(open_max+1, sizeof(WatchFile *));

	if ((kq = kqueue()) == -1)
		err(1, "cannot create kqueue");

	/* expect file list from a pipe */
	if (isatty(fileno(stdin)))
		usage();

	/* read input and populate watch list, skipping non-regular files */
	n_files = process_input(stdin, files, open_max);
	if (n_files == 0)
		errx(1, "No regular files to watch");
	if (n_files == -1)
		errx(1, "Too many files listed; the hard limit for your login"
		    " class is %u. Please consult"
		    " http://eradman.com/entrproject/limits.html", open_max);
	for (i=0; i<n_files; i++)
		watch_file(kq, files[i]);

	if (!noninteractive_opt) {
		/* Attempt to open a tty so that editors don't complain */
		ttyfd = open(_PATH_TTY, O_RDONLY);
		if (ttyfd > STDIN_FILENO) {
			if (dup2(ttyfd, STDIN_FILENO) != 0)
				warnx("can't dup2 to stdin");
			close(ttyfd);
		}

		/* remember terminal settings */
		if (tcgetattr(STDIN_FILENO, &canonical_tty) == -1)
			errx(1, "unable to get terminal attributes, use '-n' to run non-interactively");

		/* Use keyboard input as a trigger */
		EV_SET(&evSet, STDIN_FILENO, EVFILT_READ, EV_ADD, NOTE_LOWAT, 1, NULL);
		if (kevent(kq, &evSet, 1, NULL, 0, NULL) == -1)
			warnx("failed to register stdin");
	}

	watch_loop(kq, argv+argv_index);
	return 1;
}

/* Utility functions */

void
usage() {
	fprintf(stderr, "release: %s\n", RELEASE);
	fprintf(stderr, "usage: entr [-acdnprsz] [-t timeout] utility [argument [/_] ...] < filenames\n");
	exit(1);
}

void
terminate_utility() {
	int status;
	pid_t timeout_pid = 0;

	terminating = 1;

	if (child_pid > 0) {
		if (timeout_arg > 0) {
			timeout_pid = fork();
			if (timeout_pid == 0) {
				sleep(timeout_arg);
				exit(0);
			}
		}
		killpg(child_pid, SIGTERM);
		waitpid(child_pid, &status, 0);
		child_pid = 0;

		if (timeout_arg > 0) {
			if (timeout_pid > 0) {
				kill(timeout_pid, SIGKILL);
			}
		}
	}

	terminating = 0;
}

/* Callbacks */

void
handle_exit(int sig) {
	if ((!noninteractive_opt) && (termios_set))
		tcsetattr(STDIN_FILENO, TCSADRAIN, &canonical_tty);

	terminate_utility();

	if ((sig == SIGINT || sig == SIGHUP))
	    _exit(0);
	else
	    raise(sig);
}

void
proc_exit(int sig) {
	int status;
	int saved_errno = errno;

	if ((!noninteractive_opt) && (termios_set))
		tcsetattr(STDIN_FILENO, TCSADRAIN, &canonical_tty);

	if (wait(&status) != -1)
		child_status = status;

	if ((oneshot_opt == 1) && (terminating == 0)) {
		if ((shell_opt == 1) && (restart_opt == 0))
			print_child_status(child_status);

		if WIFSIGNALED(child_status)
			_exit(128 + WTERMSIG(child_status));
		else
			_exit(WEXITSTATUS(child_status));
	}
	/* restore errno so that the resuming code is unimpacted. */
	errno = saved_errno;
}

void
print_child_status(int status) {
	int len;
	char buf[2048];

	if WIFSIGNALED(status)
		len = snprintf(buf, sizeof(buf), "%s terminated by signal %d\n",
		    shell_base, WTERMSIG(status));
	else
		len = snprintf(buf, sizeof(buf), "%s returned exit code %d\n",
		    shell_base, WEXITSTATUS(status));
	write(STDOUT_FILENO, buf, len);
}

/*
 * Read lines from a file stream (normally STDIN).  Returns the number of
 * regular files to be watched or -1 if max_files is exceeded.
 */
int
process_input(FILE *file, WatchFile *files[], int max_files) {
	char buf[PATH_MAX];
	char *p, *path;
	int n_files = 0;
	struct stat sb;
	int i, matches;

	while (fgets(buf, sizeof(buf), file) != NULL) {
		if ((p = strchr(buf, '\n')) != NULL)
			*p = '\0';
		if (buf[0] == '\0')
			continue;

		if (stat(buf, &sb) == -1) {
			warnx("unable to stat '%s'", buf);
			continue;
		}
		if (S_ISREG(sb.st_mode) != 0) {
			files[n_files] = malloc(sizeof(WatchFile));
			strlcpy(files[n_files]->fn, buf, MEMBER_SIZE(WatchFile, fn));
			files[n_files]->is_dir = 0;
			files[n_files]->file_count = 0;
			files[n_files]->mode = sb.st_mode;
			files[n_files]->ino = sb.st_ino;
			n_files++;
		}
		/* also watch the directory if it's not already in the list */
		if (dirwatch_opt > 0) {
			if (S_ISDIR(sb.st_mode) != 0)
				path = &buf[0];
			else
				if ((path = dirname(buf)) == 0)
					err(1, "dirname '%s' failed", buf);
			for (matches=0, i=0; i<n_files; i++)
				if (strcmp(files[i]->fn, path) == 0) matches++;
			if (matches == 0) {
				files[n_files] = malloc(sizeof(WatchFile));
				strlcpy(files[n_files]->fn, path,
				    MEMBER_SIZE(WatchFile, fn));
				files[n_files]->is_dir = 1;
				files[n_files]->file_count = list_dir(path);
				files[n_files]->mode = sb.st_mode;
				files[n_files]->ino = sb.st_ino;
				n_files++;
			}
		}
		if (n_files+1 > max_files)
			return -1;
	}
	return n_files;
}

int list_dir(char *dir) {
	struct dirent *dp;
	DIR *dfd = opendir(dir);
	int count = 0;

	if (dfd == NULL)
		errx(1, "unable to open directory: '%s'", dir);
	while((dp = readdir(dfd)) != NULL)
		if ((dirwatch_opt == 2) || (dp->d_name[0] != '.'))
			count++;
	closedir(dfd);
	return count;
}

/*
 * Evaluate command line arguments and return an offset to the command to
 * execute.
 */
int
set_options(char *argv[]) {
	int ch;
	int argc;
	int iterations;
	const char *errstr;

	/* read arguments until we reach a command */
	argc = 1;
	while (argv[argc] != 0 && argv[argc][0] == '-') {
		ch = argv[argc][1];
		switch (ch) {
		case 't':
			argc += 2;
			break;
		default:
			argc += 1;
			break;
		 }
	}
	while ((ch = getopt(argc, argv, "acdnprszt:")) != -1) {
		switch (ch) {
		case 'a':
			aggressive_opt = 1;
			break;
		case 'c':
			clear_opt = clear_opt ? 2 : 1;
			break;
		case 'd':
			dirwatch_opt = dirwatch_opt ? 2 : 1;
			break;
		case 'n':
			noninteractive_opt = 1;
			break;
		case 'p':
			postpone_opt = 1;
			break;
		case 'r':
			restart_opt = 1;
			break;
		case 's':
			shell_opt = 1;
			break;
		case 'z':
			oneshot_opt = 1;
			break;
		case 't':
			timeout_arg = strtonum(optarg, 1, INT_MAX, &errstr);
			if (errstr != NULL)
				errx(1, "timeout %s: %s", errstr, optarg);
			break;
		default:
			usage();
		}
	}
	if (argv[optind] == 0)
		usage();
	if (timeout_arg > 0 && restart_opt == 0)
		errx(1, "timeout may only be set for restart mode");
	if ((shell_opt == 1) && (argv[optind+1] != 0))
		errx(1, "-s requires commands to be formatted as a single argument");
	return optind;
}

/*
 * Execute the program supplied on the command line. If restart was set
 * then send the child process SIGTERM and restart it.
 */
void
run_utility(char *argv[]) {
	int pid;
	int i, m;
	int ret, status;
	struct timespec delay = { 0, 1000000 };
	char **new_argv;
	char *p, *arg_buf;
	int argc;

	if (restart_opt == 1)
		terminate_utility();

	if (shell_opt == 1) {
		/* run argv[1] with a shell using the leading edge as $0 */
		argc = 4;
		arg_buf = malloc(ARG_MAX);
		new_argv = calloc(argc+1, sizeof(char *));
		realpath(leading_edge->fn, arg_buf);
		new_argv[0] = shell;
		new_argv[1] = "-c";
		new_argv[2] = argv[0];
		new_argv[3] = arg_buf;
	}
	else {
		/* clone argv on each invocation to make the implementation of more
		 * complex subsitution rules possible and easy
		 */
		for (argc=0; argv[argc]; argc++);
		arg_buf = malloc(ARG_MAX);
		new_argv = calloc(argc+1, sizeof(char *));
		for (m=0, i=0, p=arg_buf; i<argc; i++) {
			new_argv[i] = p;
			if ((m < 1) && (strcmp(argv[i], "/_")) == 0) {
				p += strlen(realpath(leading_edge->fn, p));
				m++;
			}
			else
				p += strlcpy(p, argv[i], ARG_MAX - (p - arg_buf));
			p++;
		}
	}

	pid = fork();
	if (pid == -1)
		err(1, "can't fork");

	if (pid == 0) {
		/* 2J - erase the entire display
		 * 3J - clear scrollback buffer
		 * H  - set cursor position to the default
		 */
		if (clear_opt == 1)
			printf("\033[2J\033[H");
		if (clear_opt == 2)
			printf("\033[2J\033[3J\033[H");
		fflush(stdout);

		/* Set process group so subprocess can be signaled */
		if (restart_opt == 1) {
			setpgid(0, getpid());
			close(STDIN_FILENO);
			open(_PATH_DEVNULL, O_RDONLY);
		}
		/* wait up to 1 seconds for each file to become available */
		for (i=0; i < 10; i++) {
			ret = execvp(new_argv[0], new_argv);
			if (errno == ETXTBSY)
				nanosleep(&delay, NULL);
			else break;
		}
		if (ret != 0)
			err(1, "exec %s", new_argv[0]);
	}
	child_pid = pid;

	if (restart_opt == 0 && oneshot_opt == 0) {
		if (wait(&status) != -1)
			child_status = status;

		if (shell_opt == 1)
			print_child_status(child_status);
	}

	free(arg_buf);
	free(new_argv);
}

/*
 * Wait for file to become accessible and register a kevent to watch it
 */
void
watch_file(int kq, WatchFile *file) {
	struct kevent evSet;
	int i;
	struct timespec delay = { 0, 100 * 1000000 };

	/* wait up to 1 second for file to become available */
	for (i=0; i < 10; i++) {
		#ifdef O_EVTONLY
		file->fd = open(file->fn, O_RDONLY|O_CLOEXEC|O_EVTONLY);
		#else
		file->fd = open(file->fn, O_RDONLY|O_CLOEXEC);
		#endif
		if (file->fd == -1) nanosleep(&delay, NULL);
		else break;
	}
	if (file->fd == -1) {
		warn("cannot open '%s'", file->fn);
		terminate_utility();
		exit(1);
	}

	EV_SET(&evSet, file->fd, EVFILT_VNODE, EV_ADD | EV_CLEAR, NOTE_ALL, 0,
	    file);
	if (kevent(kq, &evSet, 1, NULL, 0, NULL) == -1) {
		if (errno == ENOSPC)
			errx(1, "Unable to allocate memory for kernel queue."
			    " Please consult"
			    " http://eradman.com/entrproject/limits.html");
		else
			err(1, "failed to register VNODE event");
	}
}

/*
 * Wait for directory contents to stabilize
 */
int
compare_dir_contents(WatchFile *file) {
	int i;
	struct timespec delay = { 0, 100 * 1000000 };

	/* wait up to 0.5 seconds for file to become available */
	for (i=0; i < 5; i++) {
		if (list_dir(file->fn) == file->file_count)
			return 0;
		nanosleep(&delay, NULL);
	}
	return 1;
}

/*
 * Wait for events to and execute a command. Four major concerns are in play:
 *   leading_edge: Global reference to the first file to have changed
 *   reopen_only : Unlink or rename events which require us to spin while
 *                 waiting for the file to reappear. These must always be
 *                 processed
 *   collate_only: Changes that indicate that more events are likely to occur.
 *                 Watch for more events using a short timeout
 *   do_exec     : Delay execution until all events have been processed. Allow
 *                 the user to edit files while the utility is running without
 *                 any visible side-effects
 *   dir_modified: The number of files changed for a directory under watch
 */
void
watch_loop(int kq, char *argv[]) {
	struct kevent evSet;
	struct kevent evList[32];
	int nev;
	WatchFile *file;
	int i;
	struct timespec evTimeout = { 0, 1000000 };
	int reopen_only = !aggressive_opt;
	int collate_only = 0;
	int do_exec = 0;
	int dir_modified = 0;
	int leading_edge_set = 0;
	struct stat sb;
	char c;
	struct termios character_tty;

	leading_edge = files[0]; /* default */
	if (postpone_opt == 0)
		run_utility(argv);

	if (!noninteractive_opt) {
		/* disabling/restore line buffering and local echo */
		character_tty = canonical_tty;
		character_tty.c_lflag &= ~(ICANON|ECHO);
	}

main:
	if (!noninteractive_opt) {
		tcsetattr(STDIN_FILENO, TCSADRAIN, &character_tty);
		termios_set = 1;
	}

	if ((reopen_only == 1) || (collate_only == 1)) {
		nev = kevent(kq, NULL, 0, evList, 32, &evTimeout);
	}
	else {
		nev = kevent(kq, NULL, 0, evList, 32, NULL);
		dir_modified = 0;
	}

	if ((nev == -1) && (errno != EINTR))
		warn("kevent failed");

	/* escape for test runner */
	if ((nev == -2) && (collate_only == 0))
		return;

	for (i=0; i<nev; i++) {
		if (!noninteractive_opt && evList[i].filter == EVFILT_READ) {
			if (read(STDIN_FILENO, &c, 1) < 1) {
				EV_SET(&evSet, STDIN_FILENO, EVFILT_READ,
				    EV_DELETE, NOTE_LOWAT, 0, NULL);
				if (kevent(kq, &evSet, 1, NULL, 0, NULL) == -1)
					err(1, "failed to remove READ event");
			}
			else {
				if (c == ' ')
					do_exec = 1;
				if (c == 'q')
					kill(getpid(), SIGINT);
			}
		}
		if (evList[i].filter != EVFILT_VNODE)
			continue;

		file = (WatchFile *)evList[i].udata;
		if (file->is_dir == 1)
			dir_modified += compare_dir_contents(file);
	}
	if (!noninteractive_opt)
		tcsetattr(STDIN_FILENO, TCSADRAIN, &canonical_tty);

	collate_only = 0;
	for (i=0; i<nev; i++) {
		if (evList[i].filter != EVFILT_VNODE)
			continue;
		file = (WatchFile *)evList[i].udata;
		if (evList[i].fflags & NOTE_DELETE ||
		    evList[i].fflags & NOTE_RENAME) {
			EV_SET(&evSet, file->fd, EVFILT_VNODE, EV_DELETE,
			    NOTE_ALL, 0, file);
			if (kevent(kq, &evSet, 1, NULL, 0, NULL) == -1)
				err(1, "failed to remove VNODE event");
#if !defined(_LINUX_PORT)
			/* free file descriptor no longer monitored by kqueue */
			if ((file->fd != -1) && (close(file->fd) == -1))
				err(1, "unable to close file");
#endif
			watch_file(kq, file);
			collate_only = 1;
		}
	}
	if (reopen_only == 1) {
		reopen_only = 0;
		goto main;
	}

	for (i=0; i<nev && reopen_only == 0; i++) {

		if (evList[i].filter != EVFILT_VNODE)
			continue;
		file = (WatchFile *)evList[i].udata;
		if ((file->is_dir == 1) && (dir_modified == 0))
			continue;

		if (evList[i].fflags & NOTE_DELETE ||
		    evList[i].fflags & NOTE_WRITE  ||
		    evList[i].fflags & NOTE_RENAME ||
		    evList[i].fflags & NOTE_TRUNCATE) {
			if ((dir_modified > 0) && (restart_opt == 1))
				continue;
			do_exec = 1;
		}

		if (evList[i].fflags & NOTE_ATTRIB &&
		    S_ISREG(file->mode) != 0 && stat(file->fn, &sb) == 0) {
			if (file->mode != sb.st_mode) {
			    do_exec = 1;
			    file->mode = sb.st_mode;
			}
			if (file->ino != sb.st_ino) {
#if defined(_LINUX_PORT)
				do_exec = 1;
#endif
				file->ino = sb.st_ino;
			}
		}
		else if (evList[i].fflags & NOTE_ATTRIB)
			continue;

		if ((leading_edge_set == 0) &&
			    (file->is_dir == 0) &&
			    (do_exec == 1)) {
			leading_edge = file;
			leading_edge_set = 1;
		}

		if (getenv("EV_TRACE")) {
			fprintf(stderr, "%d/%d: fflags: 0x%x %s %o %s\n",
			    i, nev,
			    evList[i].fflags,
			    file->is_dir ? "d" : "r",
			    file->mode,
			    file->fn);
		}
	}

	if (collate_only == 1)
		goto main;
	if (do_exec == 1) {
		do_exec = 0;
		run_utility(argv);
		if (!aggressive_opt)
			reopen_only = 1;
		leading_edge_set = 0;
	}
	if (dir_modified > 0) {
		terminate_utility();
		errx(2, "directory altered");
	}

	goto main;
}
